File: LeetCodeTasks/75_sort-colors/solutions/java/solution.java
class Solution {
    public void sortColors(int[] nums) {
        int low = 0, mid = 0, high = nums.length - 1;
        while (mid <= high) {
            switch (nums[mid]) {
                case 0: // Red
                    swap(nums, low++, mid++);
                    break;
                case 1: // White
                    mid++;
                    break;
                case 2: // Blue
                    swap(nums, mid, high--);
                    break;
            }
        }
    }
    
    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}

File: LeetCodeTasks/75_sort-colors/solutions/c/solution.c
#include <stdio.h>

void sortColors(int* nums, int numsSize) {
    int low = 0, mid = 0, high = numsSize - 1;
    while (mid <= high) {
        switch (nums[mid]) {
            case 0:
                // Swap nums[low] and nums[mid]
                if (low != mid) {
                    nums[low] ^= nums[mid];
                    nums[mid] ^= nums[low];
                    nums[low] ^= nums[mid];
                }
                low++;
                mid++;
                break;
            case 1:
                mid++;
                break;
            case 2:
                // Swap nums[mid] and nums[high]
                if (mid != high) {
                    nums[mid] ^= nums[high];
                    nums[high] ^= nums[mid];
                    nums[mid] ^= nums[high];
                }
                high--;
                break;
        }
    }
}
